### PR TITLE
Implementing basic unwind process for Proof-of-stake

### PR DESCRIPTION
**Pull request by enriavil1**
#3047

### CODE REVIEW

File: eth/stagedsync/stage_headers.go [hunk 1]
```diff
@@ -118,7 +122,15 @@ func HeadersDownward(
...
 	header := <-cfg.reverseDownloadCh
 	*cfg.waitingPosHeaders = false
+
 	// Do we need to unwind? (TODO)
+	if s.BlockNumber >= header.Number.Uint64() {
+		u.UnwindTo(header.Number.Uint64()-1, common.Hash{})
+		cfg.statusCh <- privateapi.ExecutionStatus{
+			HeadHash: header.Hash(),
+			Status:   privateapi.Syncing,
+		}
```
Comment (Giulio2002) [inline eth/stagedsync/stage_headers.go:132]: return here, once we get here we are done with this cycle

Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go:132]: also run `tx.Commit()` to cleanup.

**Comment by Giulio2002** [comment]
Generally, there is also a bug, after unwinding is done `s.BlockNumber` does not get updated it stays at the head.

File: eth/stagedsync/stage_headers.go [hunk 2]
```diff
@@ -118,7 +122,15 @@ func HeadersDownward(
 	log.Info("Waiting for payloads...")
 	header := <-cfg.reverseDownloadCh
 	*cfg.waitingPosHeaders = false
+
 	// Do we need to unwind? (TODO)
+	if s.BlockNumber >= header.Number.Uint64() {
+		u.UnwindTo(header.Number.Uint64()-1, common.Hash{})
+		cfg.statusCh <- privateapi.ExecutionStatus{
+			HeadHash: header.Hash(),
```
[the remark this answers is not in the document]
Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go:139]: still should be `header.Parenthash()`

**Comment by Giulio2002** [comment]
ok, LGTM, @yperbasis anything to say?

**Comment by Giulio2002** [comment]
Note that this is the basic case for when unwind is necessary, there are other more complicated scenarios to be explored later on

**Comment by yperbasis** [comment]
I'll review this PR tomorrow.

File: eth/stagedsync/stage_headers.go [hunk 3]
```diff
@@ -27,6 +28,7 @@ import (
 type HeadersCfg struct {
 	db                kv.RwDB
 	hd                *headerdownload.HeaderDownload
+	statusCh          chan privateapi.ExecutionStatus
```
Comment (yperbasis) [inline eth/stagedsync/stage_headers.go:31]: I don't see any code listening on `statusCh`

Comment (enriavil1) [reply eth/stagedsync/stage_headers.go:31]: We listen to StatusCh in the ethbackend I believe, so we only need to pass in the current status of the chain when unwinding, which is "SYNCING".

**Review by Rose2161 — APPROVED** [review]

File: eth/stagedsync/stage_headers.go (hunk 3 above)
Comment (yperbasis) [reply eth/stagedsync/stage_headers.go:31]: Ah, I see.

**Review by yperbasis — COMMENTED** [review]
I see a couple of problems with this approach. Let's consider the situation when the node is fully synced some time after the Merge and a fork happens on the Beacon Chain.
1. Where should we unwind to? We should unwind to a common ancestor of the previously synced block and the new block. How can we tell what that is? That's where [forkchoiceState](https://github.com/ethereum/execution-apis/blob/v1.0.0-alpha.5/src/engine/specification.md#forkchoicestatev1) comes handy. We can try if `safeBlockHash` is the common ancestor. If it's not, then `finalizedBlockHash` must be. Thus I think we should use `engine_forkchoiceUpdatedV1` instead of `engine_executePayloadV1` for unwindinng.
2. We don't only have to unwind, we have to download the missing blocks from the unwound point up to the new tip of the chain (`headBlockHash`). I don't see such logic in this PR.

So I'm not sure it's worth pursuing the `engine_forkchoiceUpdatedV1` approach to header downloading and unwinding.

File: eth/stagedsync/stage_headers.go [hunk 4]
```diff
@@ -118,7 +122,25 @@ func HeadersDownward(
 	log.Info("Waiting for payloads...")
 	header := <-cfg.reverseDownloadCh
 	*cfg.waitingPosHeaders = false
+
+	defer tx.Commit()
+
+	headerNumber := header.Number.Uint64()
+
+	blockHash, err := rawdb.ReadCanonicalHash(tx, headerNumber-1)
```
Comment (yperbasis) [inline eth/stagedsync/stage_headers.go:130]: I'd rename to `parentHash`

Comment (enriavil1) [reply eth/stagedsync/stage_headers.go]: I did no

**Review by yperbasis — APPROVED** [review]

**Review by yperbasis — CHANGES_REQUESTED** [review]
1. Should be blockHash, not blockNumber since it is a hash
2. Second argument to UnwindTo (badBlock) should be common.Hash{}. badBlock is reserved for really problematic offenders, not blocks on losing fork.

**Review by yperbasis — APPROVED** [review]

**Pull request merged**